### PR TITLE
Add multi-line values support

### LINKED ISSUE #117: Library can't parse multiline values properly
Library doesn't parse multi-line variable values properly.

**Input**:

```bash
URLS="foo
bar"
```

**Expected:**

```go
map[string]string{
  "URLS": "foo\nbar",
}
```

**Got:**

```go
map[string]string{
  "URLS": "foo",
  "bar",
}
```

### PR DESCRIPTION
**Pull request by x1unix**
Fixes #117 issue with multi-line values parsing.

Had to refactor existing parser to implement this.
Theoretically, a new parser should be faster, since less iterations over source slice are required to do parsing.

The new parser passes all tests in `godotenv_test.go`, so it doesn't introduce any breaking changes.

**Upd**: If anyone needs this fix, you can a branch with fix or just use the latest tag from [my fork](github.com/x1unix/godotenv/).

**Replace approach**

```
replace (
	github.com/joho/godotenv => github.com/x1unix/godotenv v1.3.1-0.20200910042738-acd8c1e858a6
)
```

**Different package approach**

```
go get -u github.com/x1unix/godotenv@v1.4.0
```

### CODE REVIEW

**Comment by joho** [comment]
I would like to support multiline. Do you know if this behaviour is consistent with the ruby & node dotenv libraries?

**Comment by x1unix** [comment]
@joho I didn't use ruby or node so I don't know, but this is a correct behavior for environment variables in shell scripts.

**Comment by x1unix** [comment]
@joho here is a reference behavior (shell script):

image

**Comment by x1unix** [comment]
@joho any news regarding this PR?

**Comment by joho** [comment]
Sorry for the delay, I'm planning on doing a maintenance pass over all my outstanding GitHub over the Xmas break

**Comment by coolaj86** [comment]
@joho I know you're busy. Is there any sort of help that someone else could give to get this moving forward?

**Comment by x1unix** [comment]
@coolaj86 currently I'm just using my fork until this PR is "on review":

You can temporary redirect dependency to a fork in **go.mod**:

```
replace (
  github.com/joho/godotenv => github.com/x1unix/godotenv v1.3.1-0.20200910042738-acd8c1e858a6
)
```

**Comment by x1unix** [comment]
@joho any news?

**Comment by milon-bs23** [comment]
any update regarding this PR?

**Comment by x1unix** [comment]
@milon-bs23 you can switch to my branch until this PR is not merged:

**go.mod**

```
replace (
	github.com/joho/godotenv => github.com/x1unix/godotenv v1.3.1-0.20200910042738-acd8c1e858a6
)
```

P.S - maybe it worth to create a dedicated package with a different package URL ( `github.com/x1unix/godotenv` ) as more and more people started to fork this branch with this fix (example: https://github.com/compose-spec/godotenv/pull/2)

**Comment by joho** [comment]
I'm doing some manual testing on my end to ensure this change is in line with https://github.com/bkeepers/dotenv/blob/master/spec/dotenv/parser_spec.rb#L198-L234

If so, I'll merge and cut a release.

**Comment by joho** [comment]
@x1unix in testing your PR I've found a problem with windows newlines.

To give myself extra piece of mind that your multiline behavior matches the canonical ruby implementation I copied a few of their testcases over into `fixtures/quoted.env`. They work find on linux & mac, but fail the build on windows.

You can see what I've done on #156

I have some unrelated build cleanup work to do which I'll pull into another PR, but if you can get the windows linebreak issue resolved I can merge whenever.

File: parser.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+		return v == char
+	}
+}
+
+// isSpace reports whether the rune is a space character but not line break character
+//
+// this differs from unicode.IsSpace, which also applies line break as space
+func isSpace(r rune) bool {
+	switch r {
+	case '\t', '\v', '\f', '\r', ' ', 0x85, 0xA0:
```
Comment (coolaj86) [inline parser.go:202]: The comment and the code disagree. `\r` (carriage return) is generally considered to be a newline. Technically it is "return to the 0th position of the current row", but I that's for old printers.

**Comment by x1unix** [comment]
@joho thank you for the reply, I'll take a look at Windows build in a few days :)

**Comment by decentral1se** [comment]
Nice patch, works nicely but I've noticed that this breaks preserving inline comments :(

**Comment by austinsasko** [comment]
@joho what remains for this PR?

**Comment by acastro2** [comment]
Can we get this merged, please?

File: parser.go (hunk 1 above)
Comment (x1unix) [reply parser.go:202]: @coolaj86 `\r` character is included here for a Windows support. Windows has a different new line separator - `\r\n`.

**Comment by joho** [comment]
This PR is now indirectly merged via #156

I'm going to leave it on main without a numbered release for a little while (or do a beta release number) to let any regressions fall out first.

**Comment by joho** [comment]
So this is really on me for not reviewing close enough but there were a couple of problems in this PR.

The first was in the category of what I expected and was procrastinating about - essentially having undefined behaviour I hadn't written a test for that would be missed with the new parser. #204 was it.

In trying to fix it in #205, I noticed the bigger problem.

The new parser replaced the existing parser, but didn't clean out the old private functions, and not all of the tests were updated to point at the new parser. This was particularly problematic as `TestParsing` has all the odd edge cases and wasn't updated. I've updated all tests to call the newer code paths (and deleted the old stale paths) and there's a few more failing cases.

I'll try and get the tests passing again at some point this week but I've got a pretty full schedule, so if anyone involved here wanted to finish #205 off sooner I'd appreciate it.

**Comment by joho** [comment]
Managed to fix it last night, no need for follow up. thanks :v:
